### PR TITLE
FlyteRemote data context does not get used

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
@cosmicBboy found this issue when working on the unionml library.  To summarize,

Even though the `FlyteRemote` object creates a `FlyteContext`, when you use that remote object to fetch an execution, there is nothing tying that `LiteralsResolver` instance back to that context.  So when you try to download/upload anything, you end up using the default `FileAccessProvider` created at the [module layer](https://github.com/flyteorg/flytekit/blob/master/flytekit/core/data_persistence.py#L455).

To remediate, we can
* Add a context to the `LiteralsResolver` and cache the remote object's context when we get execution inputs/outputs.
  * Is this the only object that will need access to the remote object's context?  Is there anything else the user might interact with that would also require it?
* More golang style context management - allow `ctx` to be specified in the get function (but then the user has to manually use it at call-time).
* Push the context from the FlyteRemote onto the global stack in the constructor.  This way anything that uses the context will see it.  Some issues,
  * What if you create more than one remote object with different data configs?
  * What if another object, or the user manually, pushes additional FlyteContext objects onto the global stack? We do this already when we run workflows locally (inside the workflow/task `__call__` function).  Granted using a prior execution's input/outputs in a workflow declaration would be quite weird, but still something to consider.
* Do the above but lock the context somehow.

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Doing option 1 - it's the least amount of code.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: flytekit/core/data_persistence.py [hunk 1]
```diff
@@ -293,6 +293,14 @@ class FileAccessProvider(object):
     This is the class that is available through the FlyteContext and can be used for persisting data to the remote
     durable store.
     """
+    INSTANCE = 0
```
Comment (kumare3) [inline flytekit/core/data_persistence.py:296]: why?

Comment (kumare3) [reply flytekit/core/data_persistence.py:296]: is this only for debugging? if so lets delete this

File: flytekit/remote/remote.py [hunk 2]
```diff
@@ -145,7 +145,7 @@ def __init__(
         )
 
         # Save the file access object locally, but also make it available for use from the context.
-        FlyteContextManager.with_context(
+        FlyteContextManager.push_context(
```
Comment (kumare3) [inline flytekit/remote/remote.py:148]: :(

File: flytekit/core/data_persistence.py (hunk 1 above)
Comment (wild-endeavor) [reply flytekit/core/data_persistence.py:297]: will do

File: flytekit/remote/remote.py [hunk 3]
```diff
@@ -1380,11 +1382,11 @@ def _assign_inputs_and_outputs(
         """Helper for assigning synced inputs and outputs to an execution object."""
         with self.remote_context():
```
Comment (eapolinario) [inline flytekit/remote/remote.py:1383]: we don't really need to add this extra context manager now that we're specifying the context in the constructor of flyteremote, right? This applies to the other uses of `FlyteContext` in this file as well.

Comment (wild-endeavor) [reply flytekit/remote/remote.py:1383]: hmm.  i think this line is probably not needed, but other usages of the remote_context function are still required right?  For things like pulling down data and stuff, we need to use the remote instance's data config.

Comment (eapolinario) [reply flytekit/remote/remote.py:1383]: yes, for example in places like [1](https://github.com/flyteorg/flytekit/blob/master/flytekit/remote/remote.py#L429), [2](https://github.com/flyteorg/flytekit/blob/master/flytekit/remote/remote.py#L498), and [3](https://github.com/flyteorg/flytekit/blob/master/flytekit/remote/remote.py#L1002), right?

**Review by eapolinario — DISMISSED** [review]

**Review by eapolinario — APPROVED** [review]

**Review by eapolinario — APPROVED** [review]

**Review by eapolinario — APPROVED** [review]
